File: euler-ode/bin/cr3bp/plot-orbit.py
import os, json, yaml
import numpy as np
import matplotlib.pyplot as plt

from matplotlib.patches import Circle

# Configuration
with open('conf/config.json', 'r') as cli_cfg_file:
    cfg = json.load(cli_cfg_file)

if cfg['id'] != "cr3bp":
    raise ValueError('Configuration id mismatch: {}'.format(cfg['id']))

mu = cfg['params']['mu']
with open('conf/cr3bp/plot-orbit.yaml', 'r') as plot_opt_file:
    opt = yaml.load(plot_opt_file, Loader=yaml.Loader)

# Import calculated points
data_phi = np.genfromtxt(
    os.path.join(opt['src']['dir'], opt['src']['data']),
    delimiter='\t',
    skip_header=1
)

# Plot orbit
def mk_orbit_plot(x, y, name):
    fig = plt.figure(figsize=(8, 8))
    plt.title(". ".join(("Circular restricted 3-body problem", name)))
    plt.axis("equal")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.grid(True)
    plt.plot(
        x, y,
        fillstyle='none',
        marker='o',
        markersize=1.0,
        linestyle='None',
        alpha=0.25
    )
    xmin, xmax, ymin, ymax = plt.axis()
    # plt.gca().set_xlim([xmin, max(xmax, 1.05)])
    plt.text(
        xmin + 0.78*(xmax - xmin),
        ymax - 0.08*(ymax - ymin),
        "$\\mu$ = 1:{:.2f}".format(1/mu)
    )
    return fig

# Transformation to Ptolemy view
def rotate_orbit(x):
    t = x[0]
    return [t, x[1]*np.cos(t) - x[2]*np.sin(t) + mu*np.cos(t), x[1]*np.sin(t) + x[2]*np.cos(t) + mu*np.sin(t)]

# Plot orbit in corotating frame
fig = mk_orbit_plot(data_phi[:, 1], data_phi[:, 2], "Corotating frame")
body1 = Circle((-mu, 0), radius=0.02, color='darkslateblue', alpha=0.5)
body2 = Circle((1 - mu, 0), radius=0.01, color='crimson', alpha=0.5)
plt.gca().add_artist(body1)
plt.gca().add_artist(body2)
plt.show()
fig.savefig(os.path.join(opt['dst']['dir'], opt['dst']['plot']['orbit']['corotating']), dpi=150)

data_psi = np.apply_along_axis(rotate_orbit, axis=1, arr=data_phi)

# Plot orbit in Ptolemy view
fig = mk_orbit_plot(data_psi[:, 1], data_psi[:, 2], "Ptolemy view")
body1 = Circle((0, 0), radius=0.02, color='darkslateblue', alpha=0.5)
plt.gca().add_artist(body1)
plt.show()
fig.savefig(os.path.join(opt['dst']['dir'], opt['dst']['plot']['orbit']['ptolemy']), dpi=150)

File: euler-ode/octave/cr3bp/plot_orbit.m
% Create plots using prepared tab separated values

% Configuration
cfg = jsondecode(fileread('../../conf/config.json'));
if(!strcmp(cfg.id, "cr3bp"))
  beep();
  error(["Configuration id mismatch: ", cfg.id]);
endif

opt = jsondecode(fileread('plot_orbit.json'));

global mu;
global gap;
mu = cfg.params.mu;
gap = opt.dst.plot.gap;

% Import calculated points
data_phi = dlmread(fullfile(opt.src.dir, opt.src.data), "\t", 1, 0);

% Plot orbit
function hf = orbit_plot(x, y, name)

  global mu;
  global gap;
  marker_size = 0.25;

  hf = figure();
  hold on
  grid on
  axis equal
  axis([-1 - gap, 1 + gap, -1 - gap, 1 + gap]);
  xlabel "x"
  ylabel "y"
  set(gca, 'fontsize', 11);
  title(['Circular restricted 3-body problem. ', name])

  scatter(x, y, marker_size, 'r');

  glim = axis();
  text(
    glim(1) + 0.05*(glim(2) - glim(1)),
    glim(4) - 0.05*(glim(4) - glim(3)),
    sprintf("\\mu = 1 : %d", 1/mu)
  );
endfunction

% Transformation to Ptolemy view
function res = rotate_orbit(x)
  global mu;
  t = x(1);
  res = [t, x(2)*cos(t) - x(3)*sin(t) + mu*cos(t), x(2)*sin(t) + x(3)*cos(t) + mu*sin(t)];
endfunction

data_psi = cell2mat(arrayfun(
  @(k) rotate_orbit([data_phi(k, 1), data_phi(k, 2), data_phi(k, 3)]),
  (1:size(data_phi, 1)).',
  "UniformOutput",
  false
));

% Plot orbit in corotating frame
hf = orbit_plot(data_phi(:, [2]), data_phi(:, [3]), "Corotating frame");
plot([-mu], [0], 'bo', 'MarkerFaceColor', '#55A0CC', 'MarkerSize', 5);
plot(1 - mu, 0, 'bo', 'MarkerFaceColor', '#FF8C00', 'MarkerSize', 3)
print(hf, fullfile(opt.dst.dir, opt.dst.plot.corotating), "-S960,720");

% Plot orbit in Ptolemy view
hf = orbit_plot(data_psi(:, [2]), data_psi(:, [3]), "Ptolemy view");
plot([0], [0], 'bo', 'MarkerFaceColor', '#55A0CC', 'MarkerSize', 5);
tau=2*pi*(1:200)/200;
plot(cos(tau), sin(tau), 'bo', 'MarkerFaceColor', '#FF8C00', 'MarkerSize', 3)
print(hf, fullfile(opt.dst.dir, opt.dst.plot.ptolemy), "-S960,720");
